rs/registry/canister_data_provider/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let mut storable_key = vec![];
+        let version_b = self.version.to_be_bytes().to_vec();
+        let key_b = self.key.as_bytes().to_vec();
+
+        storable_key.extend_from_slice(&version_b);
+        storable_key.extend_from_slice(&key_b);
+
+        Cow::Owned(storable_key)
+    }
+
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
+        let bytes = bytes.as_ref();
+        let (version_bytes, key_bytes) = bytes.split_at(8);
+        let version = u64::from_be_bytes(version_bytes.try_into().expect("Invalid version bytes"));
+        let key = String::from_utf8(key_bytes.to_vec()).expect("Invalid UTF-8 in key");
+
+        Self { version, key }
+    }
+    const BOUND: Bound = Bound::Bounded {
+        max_size: MAX_REGISTRY_KEY_SIZE + size_of::<u64>() as u32,
+        is_fixed_size: false,
+    };
+}
+
+type Memory = VirtualMemory<DefaultMemoryImpl>;
+
+pub trait StableMemoryBorrower: Send + Sync {
+    fn with_borrow<R>(
+        f: impl FnOnce(&StableBTreeMap<StorableRegistryKey, StorableRegistryValue, Memory>) -> R,
+    ) -> R;
+    fn with_borrow_mut<R>(
+        f: impl FnOnce(&mut StableBTreeMap<StorableRegistryKey, StorableRegistryValue, Memory>) -> R,
+    ) -> R;
+}
+
+/// This registry data provider is designed to work with the `ic-registry-canister-client`
+/// in canisters, enabling the retrieval and storage of a registry copy in stable memory.
+///
+/// - If `keys_to_keep` is `Some(keys)`, only the specified `keys` will be stored in stable memory,
+///   while all other keys from the registry will be discarded.
+/// - If `keys_to_keep` is `None`, all keys from the registry will be retained in stable memory.
+pub struct CanisterDataProvider<S: StableMemoryBorrower> {
+    keys_to_keep: Option<HashSet<String>>,
+    _store: PhantomData<S>,
+}
+
+impl<S: StableMemoryBorrower> CanisterDataProvider<S> {
+    pub fn new(keys_to_retain: Option<HashSet<String>>) -> Self {
+        Self {
+            keys_to_keep: keys_to_retain,
+            _store: PhantomData,
+        }
+    }
+
+    async fn get_registry_changes_since(&self, version: u64) -> anyhow::Result<Vec<RegistryDelta>> {
+        let buff = serialize_get_changes_since_request(version)?;
+        let response = ic_cdk::api::call::call_raw(
+            Principal::from(REGISTRY_CANISTER_ID),
+            "get_changes_since",
+            buff,
+            0,
+        )
+        .await
+        .map_err(|(code, msg)| (code as i32, msg))
+        .unwrap();
+        let (registry_delta, _) = deserialize_get_changes_since_response(response)?;
+        Ok(registry_delta)
+    }
+
+    fn get_latest_version(&self) -> Option<u64> {
+        S::with_borrow(|local_registry| local_registry.last_key_value().map(|(k, _)| k.version))
+    }
+
+    fn add_deltas(&self, deltas: Vec<RegistryDelta>) -> anyhow::Result<()> {
+        for delta in deltas {
+            let string_key = std::str::from_utf8(&delta.key[..])
```

Comment:
We could switch this to `String::from_utf8_lossy` if the NNS team says that we may have non-utf8 keys?

Reply:
This is also used by the `nns_data_provider` but sure we could change both

Reply:
I don't think we have non utf8 keys